### PR TITLE
koord-descheduler: adapt k8s descheduler plugins

### LINKED ISSUE #828: [proposal] compatible with k8s descheduler configuration
**What is your proposal**:

Currently, users configure descheduler policies through the k8s descheduler configuration with version `descheduler/v1alpha1`. We need to be compatible with this configuration in koord-descheduler to facilitate smooth migration to koord-descheduler

### LINKED ISSUE #829: [proposal] Compatible with existing plugins of k8s descheduler
**What is your proposal**:

The k8s descheduler has some built-in plugins, but koord-descheduler does not support it yet. Compatibility with these plugins is necessary for smooth migration to k8s descheduler.

We can consider migrating the source code, or how to make these plugins compatible through go mod.

plugins:
- [x] #1035
- [x] #1036
- [x] #1037
- [x] #1038
- [x] #1039
- [x] #1040
- [x] #1041
- [x] #1042
- [x] #1043

### PR DESCRIPTION
**Pull request by eahydra**
### Ⅰ. Describe what this PR does

Integrate k8s descheduler plugins such as `DefaultEvictor`, `RemoveFailedPods`, `LowNodeUtilization`, etc. by introducing a layer of adapter.

In order to adapt to these plugins, `EvictionLimiter` is introduced to adapt to the k8s descheduler `Evictor` interface, so that these plugins can call the `NodeLimitExceeded` method to decide whether to stop the descheduling logic in advance.

In short, this PR solves the legacy compatibility issues with k8s descheduler. 

### Ⅱ. Does this pull request fix one issue?

fix #829 
fix #828 

### Ⅲ. Describe how to verify it

- Apply the following configuration  with koord-descheduler
- koord-descheduler will run  k8s descheduler plugins `RemoveFailedPods` and `RemoveDuplicates` per 10 seconds.

```yaml
apiVersion: descheduler/v1alpha2
kind: DeschedulerConfiguration
deschedulingInterval: 10s
dryRun: false
maxNoOfPodsToEvictPerNode: 2
maxNoOfPodsToEvictPerNamespace: 20
profiles:
  - name: koord-descheduler
    plugins:
      deschedule:
        disabled:
          - name: "*"
        enabled:
          - name: RemoveFailedPods
      balance:
        enabled:
          - name: RemoveDuplicates
      evict:
        disabled:
          - name: "*"
        enabled:
          - name: MigrationController
    pluginConfig:
      - name: RemoveFailedPods
        args:
          apiVersion: descheduler/v1alpha2
          kind: RemoveFailedPodsArgs
          reasons:
          - "test"
```

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

**Comment by eahydra** [comment]
/hold 
current PR just is PoC

File: pkg/descheduler/framework/plugins/removefailedpods/failedpods.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+const PluginName = removefailedpods.PluginName
+
+// RemoveFailedPods evicts pods in failed status phase that match the given args criteria
+type RemoveFailedPods struct {
+	framework.DeschedulePlugin
+}
+
+var _ framework.DeschedulePlugin = &RemoveFailedPods{}
+
+// New builds plugin from its arguments while passing a handle
```
Comment (eahydra) [inline pkg/descheduler/framework/plugins/removefailedpods/failedpods.go:27]: We can make the New function more general.

**Comment by LeoLiuYan** [comment]
How do we keep tracing latest feature update from `descheduler upstream`?

**Comment by eahydra** [comment]
/hold cancel
This PR is basically completed. Although some tests have been done, more scenarios need to be verified.  You can help with the Review first.  cc @hormes @FillZpp @LeoLiuYan @ZiMengSheng

**Review by ZiMengSheng — APPROVED** [review]

File: pkg/descheduler/framework/plugins/kubernetes/plugin.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+func (d *PluginDescriptor) New(args runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+	t := reflect.ValueOf(d.ArgsPrototype).Elem().Type()
+	defaultArgs := reflect.New(t).Interface().(runtime.Object)
+	d.ArgsDefaulter(defaultArgs)
+	if args == nil {
+		args = defaultArgs
+	} else {
+		unknownObj, ok := args.(*runtime.Unknown)
+		if !ok {
+			return nil, fmt.Errorf("got args of type %T, want *%sArgs", d.Name, args)
```
Comment (saintube) [inline pkg/descheduler/framework/plugins/kubernetes/plugin.go:151]: perhaps `fmt.Errof("", args, d.Name)`

Comment (eahydra) [reply pkg/descheduler/framework/plugins/kubernetes/plugin.go:151]: Yeah...I fixed it!

Comment (eahydra) [reply pkg/descheduler/framework/plugins/kubernetes/plugin.go]: done.

**Review by jasonliu747 — COMMENTED** [review]
/lgtm

**Review by FillZpp — APPROVED** [review]
/lgtm

**Pull request merged**